core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -270,48 +308,61 @@ final int getPendingStreamsCount() {
    * <p>This method <strong>must not</strong> be called concurrently with itself.
    */
   final void reprocess(@Nullable SubchannelPicker picker) {
-    ArrayList<PendingStream> toProcess;
+    ArrayList<PendingStream> toCreateRealStream;
+    ArrayList<PendingStream> toCheckCompletion;
     synchronized (lock) {
       lastPicker = picker;
       lastPickerVersion++;
-      if (picker == null || !hasPendingStreams()) {
+      if ((picker == null || !hasPendingStreams()) && !hasUncommittedStreams()) {
```

Comment:
I think changing this line is not necessary. Regardless of `hasUncommittedStreams`, if `return` then `stream.createRealStream()` or `drain()` will not be called and will not cause trouble.

Reply:
This line is because it needs to take care another case: when shutdown() is called but there are still uncommittedStreams, the shutdown path would return, then it will never shutdown. It relied on reprocess() to trigger the termination. (Similar to cancel() takes care of the last item and then trigger termination callback). This is different in shudownNow() which would take care of waiting uncommitted streams and then finalize termination.

Reply:
> hen shutdown() is called but there are still uncommittedStreams, the shutdown path would return, then it will never shutdown. It relied on reprocess() to trigger the termination.

`uncommittedStreams` can be considered as existing RPCs, so `shutdown()` should not terminate them. `uncommittedStreams` will complete transfer by themselves, and they don't rely on a second `reprocess()`.

Reply:
`shutdown()` never terminates then existing RPCs.  Do you mean we await() during shutdown()?
It seems it's hard to avoid that. Say, if there are both `pendingStreams` and `uncommittedStreams` when shutdown is called, so shutdown has to return. Then during next reprocess(), the newly created stream has not been drained , which would cause `pendingStreams` empty but `uncommittedstream` still has items in it, and we need a way to drain it. It looks reprocess() is that place, it would trigger in the next call or after idle timer.

Reply:
I think we should await() during shutdown(), because there is no "next" `reprocess()` after shutdown, so we can not rely on `reprocess()` for existing `uncommittedStreams`.

 Another way is introducing an abstract method `DelayedStream.onTransferComplete()`, and implementing `DelayedClientTransport.PendingStream.onTransferComplete()` to managed decrement of `uncommittedStreams.

Reply:
Thanks. Indeed it appears that idle timeout is permanently cancelled immediately after shutdown is called so there is no next `reprocess()`.  It looks like just `awaitTransferCompletion()` during shutdown is not enough, moreover, I believe generally we are not supposed to await() during shutdown.